### PR TITLE
Port tools/timeit to windows (AArch64, x64)

### PR DESCRIPTION
**Pull request by omjavaid**
This patch implements timeit timing tool found at tools/timeit.c for Windows.
It is required for compile time and run time calculations by the llvm-testsuite.

At this moment llvm testsuite tests can not be compiled with clang-cl MSVC combination
however we have also made required modifications to the cmake files to enable building
timeit on windows with clang-cl/cl.

In developer console with MSVC following command can be used to successfully build this code:

set CC=c:\\work\\llvm-dev\\build\\bin\\clang-cl.exe
set CXX=c:\\work\\llvm-dev\\build\\bin\\clang-cl.exe

cmake -G Ninja -DTEST_SUITE_SUBDIRS= ..\\llvm-test-suite

Phabricator Review: https://reviews.llvm.org/D153263

### CODE REVIEW

File: cmake/modules/Host.cmake [hunk 1]
```diff
@@ -1,4 +1,10 @@
-set(TEST_SUITE_HOST_CC "cc" CACHE STRING "C compiler targetting the host")
+
+if(WIN32 AND NOT TEST_SUITE_HOST_CC)
```
Comment (mstorsjo) [inline cmake/modules/Host.cmake:2]: Just a note here; `WIN32` indicates whether a cross compiler targets windows, while we probably want to know what host we're running on. A more accurate condition might be to check `CMAKE_HOST_SYSTEM_NAME` if that's what you're intending here?

Anyway, I'm not a fan of hardcoded defaults, but I think this probably is ok (and it can be overridden by the user when necessary).

File: tools/timeit.c [hunk 2]
```diff
@@ -74,57 +81,80 @@ static const char *g_target_redirect_stderr = 0;
...
+/* \brief If platform does not define pid_t type, then define it as int. */
+typedef int pid_t;
+#endif
+
 /* @name Resource Limit Variables */
 /* @{ */
 
 /* \brief If non-sentinel, the CPU time limit to set for the target. */
-static rlim_t g_target_cpu_limit = ~(rlim_t) 0;
+static rlim_t g_target_cpu_limit = ~(rlim_t)0;
```
Comment (mstorsjo) [inline tools/timeit.c:96]: These formatting changes would probably be good to leave out of the commit.

File: tools/timeit.c [hunk 3]
```diff
@@ -74,57 +81,80 @@ static const char *g_target_redirect_stderr = 0;
...
+
+#ifndef _WIN32
 static void terminate_handler(int signal) {
   /* If we are monitoring a process, kill its process group and assume we will
    * complete normally.
    */
   if (g_monitored_pid) {
-    fprintf(stderr, ("%s: error: received signal %d. "
-                     "killing monitored process(es): %s\n"),
+    fprintf(stderr,
```
Comment (mstorsjo) [inline tools/timeit.c:146]: Unnecessary formatting change.

File: tools/timeit.c [hunk 4]
```diff
@@ -74,57 +81,80 @@ static const char *g_target_redirect_stderr = 0;
...
             g_program_name, signal, g_target_program);
 
     /* Kill the process group of monitored_pid. */
     kill(-g_monitored_pid, SIGKILL);
     return;
   }
 
-  fprintf(stderr, "%s: error: received signal %d. exiting.\n",
-          g_program_name, signal);
+  fprintf(stderr, "%s: error: received signal %d. exiting.\n", g_program_name,
```
Comment (mstorsjo) [inline tools/timeit.c:156]: Unnecessary formatting change.

File: tools/timeit.c [hunk 5]
```diff
@@ -140,10 +170,92 @@ static void timeout_handler(int signal) {
...
+static FILE *windows_fopen(const char *filename, const char *mode) {
+  FILE *file;
+  if (fopen_s(&file, filename, mode))
+    return NULL;
+  return file;
+}
+#define fopen(filename, mode) windows_fopen(filename, mode)
+
+/* Process information struct global variable. */
+PROCESS_INFORMATION proc_info = {0};
```
Comment (mstorsjo) [inline tools/timeit.c:203]: As other globals in the file seems to be using the naming convention of a `g_` prefix for globals, we could probably use that here as well. And it might be good to point out that the code generally uses local variables with `pid_t` for passing the reference to a process around, but here we're cheating with global variables instead, which only allows one single child process.

File: tools/timeit.c [hunk 6]
```diff
@@ -219,7 +407,9 @@ static int monitor_child_process(pid_t pid, double start_time) {
     fprintf(fp, "%-10s %.4f\n", "sys", sys_time);
     fclose(fp);
   }
-
+#ifdef _WIN32
+  CloseHandle(h_output_file);
```
Comment (mstorsjo) [inline tools/timeit.c:411]: Shouldn't we have an `if (h_output_file != INVALID_HANDLE_VALUE)` around this?

File: cmake/modules/Host.cmake (hunk 1 above)
Comment (omjavaid) [reply cmake/modules/Host.cmake]: I have update this in new revision.

File: tools/timeit.c (hunk 2 above)
Comment (omjavaid) [reply tools/timeit.c]: fixed.

File: tools/timeit.c (hunk 5 above)
Comment (omjavaid) [reply tools/timeit.c]: proc_info is no longer a global after latest changes.

Comment (omjavaid) [reply tools/timeit.c]: Also commented about single threaded behaviour of the app in recent changes.

File: tools/timeit.c (hunk 6 above)
Comment (omjavaid) [reply tools/timeit.c]: yes. I have fixed this.

**Review by mstorsjo — APPROVED** [review]
LGTM, assuming that you've tested it on both Unix and Windows.
